core/src/mast/merger/mod.rs

```diff
@@ -0,0 +1,442 @@
+        // since the EqHash computation requires all descendants of a node
+        // to be in this index. Hence when we encounter a node in the merging forest
+        // which has descendants (Call, Loop, Split, ...), then those need to be in the
+        // indices.
+        self.node_id_by_hash
+            .entry(node_eq.mast_root)
+            .and_modify(|node_ids| node_ids.push((node_eq, new_node_id)))
+            .or_insert_with(|| vec![(node_eq, new_node_id)]);
+
+        self.hash_by_node_id.insert(new_node_id, node_eq);
+
+        Ok(())
+    }
+
+    /// This will handle two cases:
+    ///
+    /// - The existing forest contains a node with MAST root `foo` and the merging External node
+    ///   refers to `foo` and their fingerprints do not match. If their fingerprints match this is
+    ///   the general case of merging (or deduplicating to be precise) and is handled in
+    ///   `merge_nodes` already, so it is not duplicated here.
+    /// - The existing forest contains _one or more_ External nodes with a MAST root `foo` and the
+    ///   merging node's digest is `foo`.
+    ///
+    /// Importantly, this does not handle the case where the fingerprints match as that is the
+    /// general case of merging and is handled in `merge_nodes` already, so it is not duplicated
+    /// here.
+    fn merge_external_nodes(
+        &mut self,
+        previous_id: MastNodeId,
+        node_eq: &EqHash,
+        remapped_node: &MastNode,
+        node_id_remapping: &mut MastForestIdMap,
+    ) -> Result<ControlFlow<()>, MastForestError> {
+        if remapped_node.is_external() {
+            // If any non-external node exists, use it and drop the external node.
+            match self.lookup_non_external_node_by_root(node_eq) {
+                Some((_, referenced_node_id)) => {
+                    node_id_remapping.insert(previous_id, *referenced_node_id);
+                    Ok(ControlFlow::Break(()))
+                },
+                // If no replacement for the external node exists do nothing as `merge_nodes` will
+                // simply add the node to the forest.
+                None => Ok(ControlFlow::Continue(())),
+            }
+        } else {
+            // Replace all external nodes in self with the given MAST root with the non-external
+            // node from the merging forest.
+            // Any node in the existing forest that pointed to the external node will
+            // have the same MAST root due to the semantics of external nodes.
+            //
+            // By default we assume that no external node will be replaced in which case we want to
+            // `Continue`, otherwis we `Break`.
+            let mut control_flow = ControlFlow::Continue(());
+
+            for (_, external_node_id) in self.lookup_all_external_nodes_by_root(node_eq).into_iter()
+            {
+                self.mast_forest[external_node_id] = remapped_node.clone();
```

Comment:
Regarding the first issue:
I have changed the implementation now such that the merged forest never contains more than one External Node with the same MAST root. This is enforced in `merge_external_nodes`. If an external node with MAST root `foo` should be merged then it will be dropped if any node with the same MAST root already exists (external or non-external). That way it is guaranteed that when we get to the other branch we only have one possible external node with a given MAST root. And then if we replace that single external node with some other node, it'll be unique.
I think that solves that issue.

Regarding the second issue:
I agree this is not entirely in line with MastForests by construction, but if I understood @bitwalker correctly on Discord, then it is possible (and perhaps even desirable) for a MastForest to be able to violate this property, even if it usually doesn't:
> I think the question really boils down to: are there situations where we might mutate a MastForest by introducing new nodes, and end up with nodes containing children that were created after the their parent. For example, a MastForest initially is constructed with an external reference to some procedure, but we later want to merge in the code for that procedure, and convert any external references to it, to use the actual MastForest node identifier - this would be a situation where the parent nodes of the replaced external references, would have older indices than their children.

This sounds exactly like what happens here when we replace an external node with another one.

My assumption is that if we really rely on such a property somewhere then we must enforce it on the type level. Then we should disallow construction of `MastForest`s that violate this property - in other words, we need validation and/or a function to fix it. And since we don't enforce it completely I assume it's okay.

I don't see an easy way to fix this situation on top of the current merge process.
- Because we need all children of "reference nodes" (Loop, Split, Join, ...) to be processed before those reference nodes can be processed, we cannot process external nodes after all other non-external nodes, which would make it easy to enforce the property.
- I think it would have to be done in another separate DFS postorder pass over the entire forest which rebuilds the forest and remaps nodes.

Reply:
> For anything that is not a basic block, this could result in a situation where the parent has a smaller MastNodeId than its children. Not sure if this will break anything, but I wonder if we should try to preserve the property where IDs of parents are greater than IDs of their children.

Indeed the current deserialization assumes this (e.g. [here](https://github.com/0xPolygonMiden/miden-vm/blob/b2294a1e0e98af155d679dce082eb7fe3a04d9a2/core/src/mast/serialization/info.rs#L62) we assume when building a `JOIN` that both its children IDs exist in the forest already).

I believe this is the only place we make that assumption. I haven't thought it through, but indeed if we can fix deserialization to not make that assumption, then we don't have to change anything about how we do merging.

@PhilippGackstatter could you look into it? And possibly add a test that ensures that deserializing a `MastForest` with a child's ID is larger than a parent's ID. Might be worth doing in another PR too, not sure.

Reply:
Great, thank you for looking into that!
I opened https://github.com/0xPolygonMiden/miden-vm/pull/1542 to adapt the serialization which builds on top of `next` so it can be merged before this PR.

Reply:
How difficult would it be to refactor the approach to make sure we maintain the property that child indexes are always smaller than parent indexes? (no need to change anything yet, just curious if you have some ideas on this).

Reply:
I think a hard requirement for merging is a DFS traversal to process all children before parent nodes which is required by the fingerprint computation. And because of that traversal, we cannot pre- or postprocess external nodes. So we have to process external nodes during the traversal and so if we encounter an External node for which we don't yet know its replacement (if any), then we have to insert it. And if we touch that node later (by replacing it) then we might violate the id order property.

**Option 1**

However, the "we don't yet know" part above could be addressed by indexing all to-be-merged forests first by their MAST root.
We could consider changing the DFS traversal to operate on multiple forests and the iterator would keep that MAST root index and yield replacements for external nodes if they exist and thereby transparently iterate through all forests. Seems fairly complex to implement though, but I guess it's technically possible.

**Option 2**

As an aside, the merge process for `x + y` currently means that we first merge `x` into an empty forest, and then `y` into that forest. After `x` is merged into the empty forest, that merged forest actually has the "child id < parent id" property, because we did a DFS traversal and processed and inserted all children first. Put another way: Merging a single forest into an empty forest is the operation that establishes this id order property already.

So by far the easiest solution I can think of would be to do a second "merge" after the actual merge (which also requires merging the root maps since the indices may now have changed again relative to the original forest), since that will do a DFS traversal, establish the above property and remap all nodes. The biggest downside is just that it's fairly expensive to do that for a large forest given all the indices that are kept during merging. This could be handled on a high-level in `MastForest::merge`:

```rust
pub fn merge(
    forests: impl IntoIterator<Item = MastForest>,
) -> Result<(MastForest, Vec<MastForestRootMap>), MastForestError> {
    let mut root_maps = Vec::new();
    let mut merger = MastForestMerger::new();

    for forest in forests {
        root_maps.push(merger.merge(forest)?);
    }

    let merged_forest = merger.into();
    let mut merger = MastForestMerger::new();
    let post_merge_root_map = merger.merge(merged_forest)?;

    for root_map in root_maps.iter_mut() {
        root_map.merge_maps(&post_merge_root_map);
    }

    Ok((merger.into(), root_maps))
}
```

Other options I cannot think about. In particular I don't see an alternative that does this during the merge and would be more efficient. This is because of the interdependencies between nodes that need to be handled holistically for the entire forest.
- We cannot remove the `External` node and instead insert the replacement node at the end of the `nodes` vec, because that would require shifting all nodes and remapping all of them. So this does not help.
- We could keep a reverse reference mapping from all nodes that reference an external node. And if we replace that external node, then we will update all the nodes that refer to the replaced node, which might involve removing that node and reinserting it to establish the id ordering property. This sounds a bit like a DFS traversal, just in reverse and harder to implement.

Reply:
What if we do something similar to your second option but more light-weight. Basically, since we have an iterator over all forests, we do two traversals over them:

- The first traversal would only identify `External` nodes. Once a set of all external nodes is built, we can figure out if any of these are non-external roots in any of the forests (we could also build a set of roots first and then only keep track of external nodes which are also non-externals roots). If such roots are found, we copy subtrees under these roots into the merger.
- The second traversal would be done almost exactly as it is done now - but now, the merger already has all procedures which are referred by external nodes, and so any external node would be simply a duplicate.
  - This should simplify somewhat handling of external nodes during this traversal.

To given an example, let's say we have two forests `A` and `B`. There is a procedure in `B` and an external node in `A` which refers to this procedure. During the first traversal, we'd identify this procedure and insert it into the merger (i.e., copy the subtree of the procedure root into the merger). Then, as we iterate over nodes of `A` the external node would be detected as a duplicate and would be skipped. Similarly, during traversal of nodes of `B`, the already copied procedure would be skipped as it is already present in the merger.

Reply:
I refactored this now. The only viable option, afaict was to use the Multi-MastForest iterator approach. The approach you described, if I understood it correctly, fails in an edge case where an External node refers to another non-external node and that node has an external node in its children. Basically [this test case](https://github.com/0xPolygonMiden/miden-vm/blob/06cdb213722f9c9089efc8accda004decac71dff/core/src/mast/merger/tests.rs#L688-L694) was failing with that approach, but it works now with the multi iterator which is actually not too bad and has the same level of complexity as my initial implementation of your approach.

Reply:
Nice! I looked over the approach and my understanding is that we are basically doing two passes over the forests:

1. The first pass builds a map of all non-external nodes (by their MAST root).
2. The second pass actually merges the forests replacing an redundant external nodes with the actual nodes (how exactly this is done I didn't get fully yet, but I also didn't spend enough time on this).

Overall, I think this approach works - so, this thread can be resolved now.